### ISSUE TITLE
Supervisor uses image EXPOSE directive to expose container ports

### ISSUE THREAD

**Issue by pipex** [balena-os/balena-supervisor#2211]
This is different than what docker-compose does, and it can conflict with some options, for instance, for using `network_mode: service:<service name>`, that will fail with 

```
Error: Failed to apply state transition steps. (HTTP code 400) unexpected - conflicting options: port exposing and the container type network mode 
```

It is unclear yet why this configuration was made so it is not safe to remove it yet, but we might be able to add an exception for service network mode.

**Comment by klutchell** [comment]
I just got caught by this too. Is there any workaround to use `network_mode: service:<service name>` while we investigate?

**Comment by pipex** [comment]
What do you think about this PR #2212? It seems like a good compromise to me

**Comment by pipex** [comment]
No workaround really @klutchell, there is no way to undefine the EXPOSE directive on the docker image, and docker-compose configurations are additive. Depending on what you are trying to do there may be alternatives using network-mode host, but I understand that is not desirable.

**Comment by klutchell** [comment]
The workaround seems fine, but the expose instruction has been deprecated for years now and the SV shouldn't have to apply anything for containers to be reachable on the same bridge.

Perhaps it is leftover from when Docker had to take action on that?

Since then docker networks have changed so containers on the same network automatically have all ports exposed.

**Comment by pipex** [comment]
The change seems to be 5 years old, to [this commit](https://github.com/balena-os/balena-supervisor/blame/80f06adb0f76aefcd82dca30b40b0ea1c68c51c6/src/compose/service.coffee#L369). 

What do you mean the expose directive is deprecated? I though it could be used with the publish all option on docker run. 

You have a point that we don't really care about the expose directive as ports need to be manually published via `docker-compose`, so it might be OK to remove it

**Comment by klutchell** [comment]
https://docs.docker.com/engine/reference/builder/#expose

> The EXPOSE instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published.

Based on that, I don't think the SV should be acting on the expose instruction. But I can see that being a bigger change in case users were depending on it being published.

**Comment by klutchell** [comment]
Perhaps we need it for cases of Dockerfile publish, without compose, but in that case it runs in host network mode anyway IIRC

**Comment by pipex** [comment]
The supervisor doesn't publish the port defined with expose, it just exposes it for the container, is the equivalent of doing `docker run --expose` that doesn't really do much

**Comment by klutchell** [comment]
Exactly, the expose command doesn't do anything, and reading it from the image source to apply it at runtime certainly doesn't seem like it would add any value.

That said, I'm fine with the workaround in your PR until we can get some more brains on the original reasoning for the feature.

**Comment by klutchell** [comment]
It's possible it was used for Dockerfile-only pushes, but I thought those ran in host network mode IIRC.

**Comment by pipex** [comment]
The reason for exposing ports from the image when starting the container is that the supervisor is not able to distinguish between ports exposed via the Dockerfile directive or the docker-compose property so it has to treat them both the same. We are looking into possible drawbacks of ignoring the `expose` configuration in order to avoid issues with service network mode

**Comment by pipex** [comment]
Should be fixed by #2217
